### PR TITLE
MOTECH-2692: Create Common Conventions for API creation in MOTECH

### PR DESCRIPTION
**Pull request by palbecki**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

File: docs/source/development/coding_conventions.rst [hunk 1]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
 	<element id="task.sidebar.tasks-toggle-active"/>
 
+
+Common Conventions for API creation
+###################################
+
+HTTP methods and endpoints
+--------------------------
+
+We use following common pattern for endpoints:
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: this is not needed, when developer defines request mapping in rest controller, he/she provides only the end of url

File: docs/source/development/coding_conventions.rst [hunk 2]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
+
+``http://<motech-server-address>/module/{module-name}/<<path>>``
+
+|v| **Do** use following general pattern for <<path>>:
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: the following, general pattern for endpoints

File: docs/source/development/coding_conventions.rst [hunk 3]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
+``/{resource}``
+
+``/{resource}/{path-variable}``
+
+Path can take params.
+
+|v| **Do** use HTTP methods verbs:
+
+- GET for read resource by id or collection.
+- GET for read - lookup resource by id or collection.
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: lookup is a specific thing in MDS, this coding conventions should have rather general character

File: docs/source/development/coding_conventions.rst [hunk 4]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
+- PUT to update resource by id.
+- DELETE to delete resource by id.
+- POST for create new resources and other operations.
+
+|x| **Avoid** use verbs in URL if there is equivalent in HTTP methods.
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: I think that more readable will be group this 'Do' and 'Avoid'

File: docs/source/development/coding_conventions.rst [hunk 5]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
+
+Correct:
+
+``GET /shoes``
+
+Incorrect:
+
+``GET /shoe/list``
+
+|v| **Do** use lookups to define queries for resource. The right path pattern is:
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: remove fragment with lookups

File: docs/source/development/coding_conventions.rst [hunk 6]
```diff
@@ -292,6 +292,89 @@ If this is a link it should refer to where the link goes.
...
+
+``GET /lookup/shoes/byStyle/clog``
+
+Response codes
+--------------
+
+|v| **Do** use response codes:
+
+- **200** *OK* - For successful operation.
+- **201** *Created* - The operation has successful and resulted in a new resource being created.
```
Comment (kmadej) [inline docs/source/development/coding_conventions.rst]: remove 201 response code

**Pull request merged**